torchao/quantization/marlin_qqq/utils.py

```diff
@@ -0,0 +1,186 @@
+    Returns:
+        torch.Tensor: Weight tensor permuted to Marlin tiles.
+    """
+    assert q_w.shape == (size_k, size_n)
+    assert size_k % tile == 0, f"size_k = {size_k}, tile = {tile}"
+    assert size_n % tile == 0, f"size_k = {size_n}, tile = {tile}"
+
+    # Permute weights to 16x64 marlin tiles
+    q_w = q_w.reshape((size_k // tile, tile, size_n // tile, tile))
+    q_w = q_w.permute((0, 2, 1, 3))
+    q_w = q_w.reshape((size_k // tile, size_n * tile))
+
+    q_w = q_w.reshape((-1, perm.numel()))[:, perm].reshape(q_w.shape)
+
+    return q_w
+
+
+def reverse_marlin_permute_weights(
+    q_w_unpacked: torch.Tensor,
+    size_k: int,
+    size_n: int,
+    reverse_perm: torch.Tensor,
+    tile: int = const.TILE,
+) -> torch.Tensor:
+    """Reverse permute weights from 16x64 Marlin tiles.
+    Args:
+        q_w_unpacked (torch.Tensor): Unpacked quantized weights.
+        size_k (int): Number of input features.
+        size_n (int): Number of output features.
+        reverse_perm (torch.Tensor): The computed reverse permutation tensor to be applied.
+        tile (int, optional): Tile size. Defaults to `TILE`.
+    Returns:
+        torch.Tensor: Weight tensor reverse permuted from Marlin tiles.
+    """
+
+    assert (q_w_unpacked.shape[0], size_n) == (
+        size_k // tile,
+        q_w_unpacked.shape[1] // tile,
+    )
+    assert size_k % tile == 0, f"size_k = {size_k}, tile = {tile}"
+    assert size_n % tile == 0, f"size_k = {size_n}, tile = {tile}"
+
+    # Reverse permute weights to original shape
+    q_w_comp = q_w_unpacked.reshape((-1, reverse_perm.numel()))[
+        :, reverse_perm
+    ].reshape(q_w_unpacked.shape)
+    q_w_comp = q_w_comp.reshape((size_k // tile, size_n // tile, tile, tile))
+    q_w_comp = q_w_comp.permute((0, 2, 1, 3))
+    q_w_comp = q_w_comp.reshape((size_k, size_n))
+
+    return q_w_comp
+
+
+# NOTE(HandH1998): QQQ employs different perms for per-group and per-channel weight quantization. # noqa: E501
+def get_qqq_weight_perm(num_bits: int, quant_type: str) -> torch.Tensor:
+    """Precompute permutations for the marlin weight shuffling.
+
+    Args:
+        num_bits (int): Number of bits to pack.
+        quant_type (str): The weight quantization type: per-group or per-channel.
+    Returns:
+        torch.Tensor: The weight permutation tensor.
+    """
+    perm_list: List[int] = []
+    for i in range(32):
+        perm1: List[int] = []
+        col = i // 4
+        for block in [0, 1]:
+            for row in [
+                4 * (i % 4),
+                4 * (i % 4) + 1,
+                4 * (i % 4) + 2,
+                4 * (i % 4) + 3,
+            ]:
+                perm1.append(16 * row + col + 8 * block)
+        for j in range(4):
+            perm_list.extend([p + 256 * j for p in perm1])
+
+    perm = numpy.array(perm_list)
+
+    assert quant_type in ["per-channel", "per-group"], "not supported quantization type"
```

Comment:
quant_type seems like an enum would be more appropriate

Reply:
we have https://github.com/pytorch/ao/blob/main/torchao/quantization/granularity.py btw

Reply:
I have modified the `quant_type` to `granularity` in  https://github.com/pytorch/ao/blob/main/torchao/quantization/granularity.py.